core/diagnostics.rs

```diff
@@ -0,0 +1,871 @@
+  pub script_name: String,
+  pub function_name: String,
+  pub is_eval: bool,
+  pub is_constructor: bool,
+  pub is_wasm: bool,
+}
+
+impl DenoDiagnosticFrame {
+  // TODO Maybe use serde_derive?
+  fn from_json_value(v: &serde_json::Value) -> Option<Self> {
+    if !v.is_object() {
+      return None;
+    }
+    let obj = v.as_object().unwrap();
+
+    let line_v = &obj["line"];
+    if !line_v.is_u64() {
+      return None;
+    }
+    let line = line_v.as_u64().unwrap() as i64;
+
+    let column_v = &obj["column"];
+    if !column_v.is_u64() {
+      return None;
+    }
+    let column = column_v.as_u64().unwrap() as i64;
+
+    let script_name_v = &obj["scriptName"];
+    if !script_name_v.is_string() {
+      return None;
+    }
+    let script_name = String::from(script_name_v.as_str().unwrap());
+
+    // Optional fields. See EncodeExceptionAsJSON() in libdeno.
+    // Sometimes V8 doesn't provide all the frame information.
+
+    let mut function_name = String::from(""); // default
+    if obj.contains_key("functionName") {
+      let function_name_v = &obj["functionName"];
+      if function_name_v.is_string() {
+        function_name = String::from(function_name_v.as_str().unwrap());
+      }
+    }
+
+    let mut is_eval = false; // default
+    if obj.contains_key("isEval") {
+      let is_eval_v = &obj["isEval"];
+      if is_eval_v.is_boolean() {
+        is_eval = is_eval_v.as_bool().unwrap();
+      }
+    }
+
+    let mut is_constructor = false; // default
+    if obj.contains_key("isConstructor") {
+      let is_constructor_v = &obj["isConstructor"];
+      if is_constructor_v.is_boolean() {
+        is_constructor = is_constructor_v.as_bool().unwrap();
+      }
+    }
+
+    let mut is_wasm = false; // default
+    if obj.contains_key("isWasm") {
+      let is_wasm_v = &obj["isWasm"];
+      if is_wasm_v.is_boolean() {
+        is_wasm = is_wasm_v.as_bool().unwrap();
+      }
+    }
+
+    Some(Self {
+      line: line - 1,
+      column: column - 1,
+      script_name,
+      function_name,
+      is_eval,
+      is_constructor,
+      is_wasm,
+    })
+  }
+}
+
+impl fmt::Display for DenoDiagnosticFrame {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    let function_name = self.function_name.clone();
+    // Note when we print to string, we change from 0-indexed to 1-indexed.
+    let line = (1 + self.line).to_string();
+    let column = (1 + self.column).to_string();
+    let script_name = self.script_name.to_string();
+    let script_line_column = format!("{}:{}:{}", script_name, line, column);
+
+    if !self.function_name.is_empty() {
+      write!(f, "    at {} ({})", function_name, script_line_column)
+    } else if self.is_eval {
+      write!(f, "    at eval ({})", script_line_column)
+    } else {
+      write!(f, "    at {}", script_line_column)
+    }
+  }
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct DenoDiagnostic {
+  pub message: String,
+  pub diagnostics: Option<Vec<DenoDiagnostic>>,
```

Comment:
This seems to define a tree of diagnostics - because each Diagnostic and have many sub Diagnostics, which themselves could have sub diagnostics. Is this intended? I would imagine there's rather a list of diagnostics.

Reply:
This is going to take a bit of an explanation.  In TypeScript a `Diagnostic.message` can be either `string | DiagnosticMessageChain`.  So it is a diagnostic chain.  Having variable types in structures is, I found out, a Rust anti-pattern and so I split it so `message` was always `string` and transformed the diagnostic chain into an array/vector.  The `DiagnosticMessageChain` is actually different than `Diagnostic` but it was easy to map it and not have to have a distinct structure.

Also, TypeScript returns an array of diagnostics for any piece of code.  But when I tried to return something that mapped was an Error containing a vector in Rust, I realised that is really difficult, and so I then had to implement a chain of diagnostics (in which I learned about boxing to be able to deal with Rust not liking to know the size of something when dealing with recursion).

So upon reflection, maybe implementing the `DiagnosticChain` structure and change the value to a boxed chain instead of an array/vector might be better and more rust like (and clearer what is going on).

Reply:
Yeah, if it's possible I would prefer introducing `DiagnosticChain`. For the reference of readers not familiar the TypeScript interface is:
```ts
    interface DiagnosticMessageChain {
        messageText: string;
        category: DiagnosticCategory;
        code: number;
        next?: DiagnosticMessageChain;
    }
```
I think you can do something like this:
```rs
struct DiagnosticMessage {
  pub message_text: String,
  pub category: DiagnosticCategory,
  code: i32,
}

struct Diagnostic {
  pub message: String,
  pub diagnostics: Option<Vec<DiagnosticMessage>>,
  pub related_information: Option<Vec</* not sure about this */>>,
  pub source_line: Option<String>,
  pub line_number: Option<i64>,
  pub script_resource_name: Option<String>,
  pub start_position: Option<i64>,
  pub end_position: Option<i64>,
  pub category: DenoDiagnosticCategory,
  pub code: Option<i64>,
  pub source: DenoDiagnosticSources,
  pub start_column: Option<i64>,
  pub end_column: Option<i64>,
  pub frames: Option<Vec<DenoDiagnosticFrame>>,
  pub next: Option<Box<DenoDiagnostic>>,
}
```

Reply:
I already had most of the first part done, but used `DiagnosticMessageChain`.

As for the second part. It is already working and the Rust code only required the next value to be boxed to deal with the recursion. I think it works well, and all JS errors are a single anyways, so the TS ones are the exception to the rule. Then again, we could put the source in the diagnostic shell and simple have a vector of items. Hmmmm... Ok I talked myself into it.